### PR TITLE
Client Request Cached Methods List from JIT Server

### PR DESCRIPTION
**Pull request by luke-li-2003**
Added an option that makes the client request a list of cached methods from the server on bootstrap, and set the count for those methods as scount to improve rampup

Depends on https://github.com/eclipse/omr/pull/7458

### CODE REVIEW

**Comment by luke-li-2003** [comment]
@mpirvu

**Review by marius-pirvu — COMMENTED** [review]
I provided a partial review of the code.

File: runtime/compiler/control/rossa.cpp [hunk 1]
```diff
@@ -1917,6 +1920,56 @@ onLoadInternal(
...
+                                     cachedMethods.size());
+
+   std::unordered_set<std::string> *serverAOTMethodSet =
+      new (PERSISTENT_NEW) std::unordered_set<std::string>(cachedMethods.begin(),
+                                                            cachedMethods.end());
+
+   client->~ClientStream();
+   TR_Memory::jitPersistentFree(client);
+
+   vm->serverAOTMethodSet = (UDATA) serverAOTMethodSet;
```
[the remark this answers is not in the document]
Comment (luke-li-2003) [reply runtime/compiler/control/rossa.cpp:1966]: Is there an example on how to do this?

File: runtime/compiler/control/rossa.cpp [hunk 2]
```diff
@@ -1917,6 +1920,56 @@ onLoadInternal(
...
+   {
+   J9JITConfig *jitConfig = (J9JITConfig *) entryarg;
+   J9JavaVM *vm = jitConfig->javaVM;
+   TR::CompilationInfo * compInfo = TR::CompilationInfo::get(jitConfig);
+   TR::PersistentInfo *persistentInfo = compInfo->getPersistentInfo();
+
+   j9thread_t osThread = (j9thread_t) vm->serverAOTQueryThread;
+   J9VMThread *vmThread = NULL;
+
+   int rc = vm->internalVMFunctions->internalAttachCurrentThread(vm, &vmThread, NULL,
```
[the remark this answers is not in the document]
Comment (luke-li-2003) [reply runtime/compiler/control/rossa.cpp:1934]: How do we wait for the attach? Here I simply referenced the code from `listenerThreadProc`.

**Comment by luke-li-2003** [comment]
Forgot to force-push, should be fine now

File: runtime/compiler/control/rossa.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (marius-pirvu) [reply runtime/compiler/control/rossa.cpp]: FLUSH_MEMORY(TR::Compiler->target.isSMP());

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 3]
```diff
@@ -425,6 +428,42 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
...
+      // Construct a signature
+      J9UTF8 *className;
+      J9UTF8 *name;
+      J9UTF8 *signature;
+      getClassNameSignatureFromMethod(method, className, name, signature);
+      // SigLen calculation as used in TRJ9VMBase::printTruncatedSignature
+      int32_t sigLen = J9UTF8_LENGTH(className) + J9UTF8_LENGTH(name) +
+         J9UTF8_LENGTH(signature) + 2;
+
+      char *sigC = new char[sizeof(char) * (sigLen + 1)];
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:447]: OpenJ9 has its own memory management that provides accounting, so using the system allocator is not a good choice. On the other hand we want to minimize the usage of persistent allocator because of the overhead and fragmentation it can create. Here's an example on how this achieved in other parts of the code: https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/control/HookedByTheJit.cpp#L180
Basically we use the stack for most methods, and for the few very long signatures we use the persistent allocator. In this case I think that we can not set a low count for methods that have a very long name. These should be very few.

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 4]
```diff
@@ -425,6 +428,42 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
...
+      sigLen = sprintf(sigC, "%.*s.%.*s%.*s",
+                        J9UTF8_LENGTH(className), utf8Data(className),
+                        J9UTF8_LENGTH(name), utf8Data(name),
+                        J9UTF8_LENGTH(signature), utf8Data(signature));
+
+      // contains
+      methodCachedAtServer =
+         (serverAOTMethodSet->find(std::string(sigC)) == serverAOTMethodSet->end());
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:455]: I think that == needs to be changed to !=

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 5]
```diff
@@ -475,6 +514,20 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
                      TR::Options::getHighCodeCacheOccupancyBCount() :
                      TR::Options::getHighCodeCacheOccupancyCount();
          }
+#if defined(J9VM_OPT_JITSERVER)
+      else if (methodCachedAtServer) // Higher priority than SCC
+         {
+         int32_t scount = optionsAOT->getInitialSCount();
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:520]: Please check: if local shared classes is not used, `optionsAOT` may not exist.

Comment (marius-pirvu) [reply runtime/compiler/control/HookedByTheJit.cpp:520]: You may use `countInOptionSet || TR::Options::getCountsAreProvidedByUser()` to check whether the use has specified counts.

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 6]
```diff
@@ -1101,6 +1108,58 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
       stream->writeError(compilationLowPhysicalMemory, (uint64_t) computeServerMemoryState(getCompilationInfo()));
       abortCompilation = true;
       }
+   catch (const JITServer::StreamAotCacheMapRequest &e)
+      {
+      const std::string aotCacheName = e.getCacheName();
```
Comment (marius-pirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:1113]: Let's make this a reference since that's what `getCacheName()` returns.

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 7]
```diff
@@ -475,6 +515,24 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
                      TR::Options::getHighCodeCacheOccupancyBCount() :
                      TR::Options::getHighCodeCacheOccupancyCount();
          }
+#if defined(J9VM_OPT_JITSERVER)
+      else if (methodCachedAtServer) // Higher priority than SCC
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:519]: I think that this code should not have higher priority than the case when a method is in local SCC, because the JIT will load the method from local SCC if available (it's just faster than the remote AOT cache)

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 8]
```diff
@@ -1101,6 +1108,58 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
...
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         {
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            "compThreadID=%d handling request for AOT cache %s method list",
+            getCompThreadId(), aotCacheName.c_str());
+         }
+
+      auto aotCacheMap = compInfo->getJITServerAOTCacheMap();
+      bool pending = false;
+      auto aotCache = aotCacheMap->get(aotCacheName, 0, pending);
```
Comment (marius-pirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:1124]: Ask Christian about the interaction of this code with AOT cache persistence.
`@param pending Output flag indicating that a cache load operation from file is in progress, but not read yet`
Can we just ignore the `pending` parameter?

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 9]
```diff
@@ -1101,6 +1108,58 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
...
+         {
+            {
+            OMR::CriticalSection cs(cachedMethodMonitor);
+
+            for (;cachedAOTMethod != NULL;
+               cachedAOTMethod = cachedAOTMethod->getNextRecord())
+               {
+               const SerializedAOTMethod &serializedAOTMethod = cachedAOTMethod->data();
+               std::string sig = std::string(serializedAOTMethod.signature());
+               methodSignaturesV.push_back(sig);
```
Comment (marius-pirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:1139]: Suggestion
`methodSignaturesV.push_back(std::string(serializedAOTMethod.signature());:`

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 10]
```diff
@@ -1101,6 +1108,58 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
...
+      auto aotCache = aotCacheMap->get(aotCacheName, 0, pending);
+      auto cachedAOTMethod = aotCache->getCachedMethodHead();
+      auto cachedMethodMonitor = aotCache->getCachedMethodMonitor();
+
+      std::vector<std::string> methodSignaturesV;
+      try
+         {
+            {
+            OMR::CriticalSection cs(cachedMethodMonitor);
+
```
Comment (marius-pirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:1133]: Suggestion:
```
size_t numAOTMethods = _cachedMethodMap.size(); // Get number of methods in AOT cache (needs an accessor)
methodSignaturesV.reserve(numAOTMethods);
```

**Review by marius-pirvu — COMMENTED** [review]
Since the IDs of the various network messages has changed, we need to increment the minor version of the server.

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 11]
```diff
@@ -4181,6 +4239,9 @@ void JitShutdown(J9JITConfig * jitConfig)
 
    TR::CompilationInfo * compInfo = TR::CompilationInfo::get(jitConfig);
 
+   if (jitConfig->serverAOTMethodSet)
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:4242]: We need to destruct the set. Calling just jitPersistentFree will only deallocate the memory used for the set itself, but not the entries added to the set.

File: runtime/compiler/control/JITServerCompilationThread.cpp (hunk 10 above)
Comment (marius-pirvu) [reply runtime/compiler/control/JITServerCompilationThread.cpp:1133]: Reading the size of the map needs to be done after we obtain the lock.

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 12]
```diff
@@ -1101,6 +1108,70 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
...
+                  {
+                  const SerializedAOTMethod &serializedAOTMethod = cachedAOTMethod->data();
+                  methodSignaturesV.push_back(std::string(serializedAOTMethod.signature()));
+                  }
+               }
+            }
+         catch (const std::bad_alloc &e)
+            {
+            if (TR::Options::isAnyVerboseOptionSet(TR_VerboseJITServer,
+                                                   TR_VerboseCompilationDispatch))
```
Comment (marius-pirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:1150]: I don't see a reason for printing a message into vlog under `TR_VerboseCompilationDispatch`

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 13]
```diff
@@ -628,6 +669,24 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
             }
 #endif // defined(J9VM_INTERP_AOT_COMPILE_SUPPORT) && defined(J9VM_OPT_SHARED_CLASSES) && (defined(TR_HOST_X86) || defined(TR_HOST_POWER) || defined(TR_HOST_S390) || defined(TR_HOST_ARM) || defined(TR_HOST_ARM64))
          } // if (TR::Options::sharedClassCache())
+#if defined(J9VM_OPT_JITSERVER)
+      else if (methodCachedAtServer) // Not cached in SCC but the server
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:673]: We would like to set a low count either when SCC/AOT is not set, or when it is set but the method is not in local SCC.
We may have to use the code below in more than one place. Maybe define it in a function?

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 14]
```diff
@@ -628,6 +690,15 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
             }
 #endif // defined(J9VM_INTERP_AOT_COMPILE_SUPPORT) && defined(J9VM_OPT_SHARED_CLASSES) && (defined(TR_HOST_X86) || defined(TR_HOST_POWER) || defined(TR_HOST_S390) || defined(TR_HOST_ARM) || defined(TR_HOST_ARM64))
          } // if (TR::Options::sharedClassCache())
+
+#if defined(J9VM_OPT_JITSERVER)
```
Comment (marius-pirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:694]: We have to do the same for the case: "// AOT Body not in SCC, so scount was not set". There are paths that set some value to `count` in that case.
Similarly, we have to cover the case: "// ROM class not in shared class cache (no space or ineligible - maybe because of implicit SCC)"

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 15]
```diff
@@ -1101,6 +1108,69 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
...
+         }
+
+      auto aotCacheMap = compInfo->getJITServerAOTCacheMap();
+      bool pending = false;
+      auto aotCache = aotCacheMap->get(aotCacheName, 0, pending);
+
+      std::vector<std::string> methodSignaturesV;
+      if (aotCache)
+         {
+         auto cachedAOTMethod = aotCache->getCachedMethodHead();
```
Comment (marius-pirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:1129]: Lines 1129 and 1132 needs to stay inside the critical section `OMR::CriticalSection cs(cachedMethodMonitor);`

File: runtime/compiler/control/rossa.cpp [hunk 16]
```diff
@@ -1917,6 +1920,83 @@ onLoadInternal(
...
+   try
+      {
+      JITServer::ClientStream *client = new (PERSISTENT_NEW) JITServer::ClientStream(persistentInfo);
+      client->write(JITServer::MessageType::AOTCacheMap_request,
+                     persistentInfo->getJITServerAOTCacheName());
+
+      client->read();
+      auto result = client->getRecvData<std::vector<std::string>>();
+
+      std::vector<std::string> cachedMethods = std::get<0>(result);
```
Comment (marius-pirvu) [inline runtime/compiler/control/rossa.cpp:1953]: We should use a reference here: ` std::vector<std::string> &cachedMethods = std::get<0>(result);`

File: runtime/compiler/control/rossa.cpp [hunk 17]
```diff
@@ -2184,6 +2264,43 @@ aboutToBootstrap(J9JavaVM * javaVM, J9JITConfig * jitConfig)
    UT_MODULE_LOADED(J9_UTINTERFACE_FROM_VM(javaVM));
    Trc_JIT_VMInitStages_Event1(curThread);
    Trc_JIT_portableSharedCache_enabled_or_disabled(curThread, J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE) ? 1 : 0);
+
+#if defined(J9VM_OPT_JITSERVER)
+   jitConfig->serverAOTMethodSet = NULL;
+   if (TR::Options::getCmdLineOptions()->getOption(TR_RequestJITServerCachedMethods))
```
Comment (marius-pirvu) [inline runtime/compiler/control/rossa.cpp:2270]: There should be some code, possible in the late option processing phase, where this option is turned off if the client does not use the AOT cache.

**Comment by mpirvu** [comment]
Please rebase to resolve conflicts.

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 18]
```diff
@@ -628,6 +690,21 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
...
+      // If the method is cached at the server side, set the count to the lower number
+      // regardless of whether the count has been set before
+      if (methodCachedAtServer)
```
Comment (mpirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:697]: I think we should also test `TR::Options::getCountsAreProvidedByUser()`.
Let's say that the use specified -Xaot:count=300, then we should use that 300 value regardless of whether the method is cached at the server or not.
Then we can simplify `calculateCountForMethodCachedAtServer()` to be just `optionsAOT->getInitialSCount()`

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 19]
```diff
@@ -425,6 +449,44 @@ static void jitHookInitializeSendTarget(J9HookInterface * * hook, UDATA eventNum
          }
       }
 
+#if defined(J9VM_OPT_JITSERVER)
```
[the remark this answers is not in the document]
Comment (mpirvu) [reply runtime/compiler/control/HookedByTheJit.cpp:452]: Maybe test here `TR::Options::getCountsAreProvidedByUser()` and avoid finding whether the method is cached at the server.

File: runtime/compiler/control/rossa.cpp [hunk 20]
```diff
@@ -2191,6 +2271,48 @@ aboutToBootstrap(J9JavaVM * javaVM, J9JITConfig * jitConfig)
...
+      TR::Options::getCmdLineOptions()->setOption(TR_RequestJITServerCachedMethods, false);
+      }
+
+   jitConfig->serverAOTMethodSet = NULL;
+   if (TR::Options::getCmdLineOptions()->getOption(TR_RequestJITServerCachedMethods))
+      {
+      // Ask the server for its cached methods
+      if (compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::CLIENT)
+         {
+         if (JITServerHelpers::isServerAvailable())
```
[the remark this answers is not in the document]
Comment (luke-li-2003) [reply runtime/compiler/control/rossa.cpp:2287]: There is no explicit error thrown, `JITServerHelpers::isServerAvailable()` indeed does not catch the error, but `catch (const JITServer::StreamFailure &e)` in the thread does.

**Review by mpirvu — COMMENTED** [review]
LGTM.
Could you please obtain some rampup curves to see whether we do better by using this feature. Use AcmeAir, preferably in containers, if not then outside containers without a SCC.

**Comment by luke-li-2003** [comment]
Here are the rampup curves using AcmeAirEE8 outside containers (without SCC) for this branch and the master branch.

image

**Comment by luke-li-2003** [comment]
Also the result on my local machine with jdk17 (the previous curves are of jdk11)

image

**Comment by mpirvu** [comment]
jenkins compile all jdk17

**Comment by mpirvu** [comment]
jenkins test sanity xlinux zlinux jdk17

**Comment by mpirvu** [comment]
jenkins test sanity xlinuxjit zlinuxjit jdk17

**Comment by mpirvu** [comment]
jenkins test sanity xlinuxjit,zlinuxjit jdk17

**Comment by mpirvu** [comment]
Tests have passed, but there are conflicts which need to be resolved

**Comment by mpirvu** [comment]
jenkins test sanity zlinuxjit jdk21

**Comment by luke-li-2003** [comment]
Running without the option `Xjit:requestJITServerCachedMethods`, the performance difference is negligible compared to the master branch build. 

```
Results for jdk: /team/lukeli/jdk11Master and opts: -Xmx256m -Xms256m -Xjit:verbose={JITServer},verbose={counts},     verbose={compilePerformance},vlog=/team/lukeli/logs/vlog.txt -Xshareclasses:none
First run is a cold run and is not included in the stats
Run 0    3581.3  4683.1 Avg= 4683.1  RSS=    231 MB  CompCPU= 34.1 sec  Startup= 6042 ms
Run 1    7908.7  4771.1 Avg= 4771.1  RSS=    273 MB  CompCPU= 36.3 sec  Startup= 5965 ms
Run 2    3584.2  4713.0 Avg= 4713.0  RSS=    228 MB  CompCPU= 32.6 sec  Startup= 5898 ms
Run 3    6862.2  4757.3 Avg= 4757.3  RSS=    259 MB  CompCPU= 36.9 sec  Startup= 5994 ms
Run 4    3586.6  4716.2 Avg= 4716.2  RSS=    229 MB  CompCPU= 32.9 sec  Startup= 6215 ms
Run 5    6302.6  4734.1 Avg= 4734.1  RSS=    255 MB  CompCPU= 36.4 sec  Startup= 5936 ms
Run 6    6648.8  4758.9 Avg= 4758.9  RSS=    258 MB  CompCPU= 35.3 sec  Startup= 5807 ms
Run 7    5921.6  4729.7 Avg= 4729.7  RSS=    258 MB  CompCPU= 35.2 sec  Startup= 6244 ms 
Run 8    6131.3  4575.7 Avg= 4575.7  RSS=    262 MB  CompCPU= 36.1 sec  Startup= 5915 ms
Run 9    6278.2  4828.4 Avg= 4828.4  RSS=    264 MB  CompCPU= 36.3 sec  Startup= 5920 ms
Avg:     5913.8  4731.6 Thr= 4731.6  RSS=    254 MB  CompCPU= 35.3 sec  Startup= 5988 ms
Throughput stats: Avg= 4731.6  StdDev=   68.2  Min= 4575.7  Max= 4828.4  Max/Min=   6% CI95=    1.1% numSamples=  9   Footprint stats:  Avg=  254.0  StdDev=   15.4  Min=  228.1  Max=  273.0  Max/Min=  20% CI95=    4.7% numSamples=  9
Comp CPU stats:   Avg=   35.3  StdDev=    1.6  Min=   32.6  Max=   36.9  Max/Min=  13% CI95=    3.4% numSamples=  9   StartupTime stats:Avg= 5988.2  StdDev=  146.3  Min= 5807.0  Max= 6244.0  Max/Min=   8% CI95=    1.9% numSamples=  9
```

```
Results for jdk: /team/lukeli/jdk11GetServer and opts: -Xmx256m -Xms256m -Xjit:verbose={JITServer},verbose={counts},  verbose={compilePerformance},vlog=/team/lukeli/logs/vlog.txt -Xshareclasses:none
First run is a cold run and is not included in the stats
Run 0    6589.8  4673.6 Avg= 4673.6  RSS=    261 MB  CompCPU= 35.8 sec  Startup= 5926 ms
Run 1    3513.3  4698.9 Avg= 4698.9  RSS=    227 MB  CompCPU= 34.4 sec  Startup= 6118 ms
Run 2    7301.8  4756.4 Avg= 4756.4  RSS=    272 MB  CompCPU= 40.6 sec  Startup= 6834 ms
Run 3    6317.3  4791.8 Avg= 4791.8  RSS=    433 MB  CompCPU= 35.9 sec  Startup= 6052 ms
Run 4    3587.1  4726.7 Avg= 4726.7  RSS=    228 MB  CompCPU= 32.0 sec  Startup= 5905 ms
Run 5    6276.0  4824.8 Avg= 4824.8  RSS=    262 MB  CompCPU= 36.3 sec  Startup= 5953 ms
Run 6    7234.2  4737.9 Avg= 4737.9  RSS=    269 MB  CompCPU= 34.7 sec  Startup= 5975 ms
Run 7    3575.9  4647.6 Avg= 4647.6  RSS=    226 MB  CompCPU= 32.9 sec  Startup= 5845 ms
Run 8    3549.0  4556.8 Avg= 4556.8  RSS=    227 MB  CompCPU= 33.7 sec  Startup= 5982 ms
Run 9    3500.8  4549.4 Avg= 4549.4  RSS=    229 MB  CompCPU= 34.6 sec  Startup= 6276 ms
Avg:     4983.9  4698.9 Thr= 4698.9  RSS=    264 MB  CompCPU= 35.0 sec  Startup= 6104 ms
Throughput stats: Avg= 4698.9  StdDev=   97.0  Min= 4549.4  Max= 4824.8  Max/Min=   6% CI95=    1.6% numSamples=  9
Footprint stats:  Avg=  263.7  StdDev=   66.4  Min=  225.9  Max=  432.9  Max/Min=  92% CI95=   19.4% numSamples=  9
Comp CPU stats:   Avg=   35.0  StdDev=    2.5  Min=   32.0  Max=   40.6  Max/Min=  27% CI95=    5.5% numSamples=  9
StartupTime stats:Avg= 6104.4  StdDev=  301.3  Min= 5845.0  Max= 6834.0  Max/Min=  17% CI95=    3.8% numSamples=  9
```

**Comment by mpirvu** [comment]
> Running without the option Xjit:requestJITServerCachedMethods, the performance difference is negligible compared to the master branch build.

Are these runs done in client mode? I don't see `-XX:+UseJITServer` in the list of options.

Also, the results for `jdk11Master` are strange in that the the throughput is much larger for the first 5 minute load period than for the second one.

**Comment by luke-li-2003** [comment]
> Are these runs done in client mode? 

No, that data set was run without the JITServer

**Comment by mpirvu** [comment]
I am willing to merge this code even though peak throughput is affected. Because of this, the new functionality will not be enabled by default. However, I wanted to make sure that the resulting JITServer configuration is not doing worse in the default case than the current master branch. Hence, the comparison should use JITServer and AOT cache, just don't provide the option that enables the new code.

**Comment by mpirvu** [comment]
jenkins test sanity xlinuxjit jdk21

**Comment by luke-li-2003** [comment]
Running as the clients, the performance differences are negligible.

```
Results for jdk: /team/lukeli/jdk11GetServer and opts: -Xmx256m -Xms256m -Xjit:verbose={JITServer},verbose={counts},verbose={compilePerformance},vlog=/team/lukeli/logs/vlog.txt -XX:+UseJITServer -XX:+JITServerUseAOTCache -Xshareclasses:none
Run 0	 3944.3	 4629.5	Avg= 4629.5  RSS=    239 MB  CompCPU=  7.9 sec  Startup= 5242 ms
Run 1	 3817.4	 4441.8	Avg= 4441.8  RSS=    237 MB  CompCPU=  5.6 sec  Startup= 4419 ms
Run 2	 3997.8	 4571.5	Avg= 4571.5  RSS=    236 MB  CompCPU=  5.7 sec  Startup= 4442 ms
Run 3	 3980.1	 4650.3	Avg= 4650.3  RSS=    231 MB  CompCPU=  5.6 sec  Startup= 4435 ms
Run 4	 3998.5	 4304.3	Avg= 4304.3  RSS=    236 MB  CompCPU=  5.4 sec  Startup= 4392 ms
Run 5	 4143.3	 4634.1	Avg= 4634.1  RSS=    232 MB  CompCPU=  5.4 sec  Startup= 4454 ms
Run 6	 3975.4	 4429.6	Avg= 4429.6  RSS=    234 MB  CompCPU=  5.5 sec  Startup= 4411 ms
Run 7	 3967.6	 4607.4	Avg= 4607.4  RSS=    234 MB  CompCPU=  5.5 sec  Startup= 4446 ms
Run 8	 3956.7	 4270.0	Avg= 4270.0  RSS=    230 MB  CompCPU=  5.4 sec  Startup= 4395 ms
Run 9	 3978.2	 4462.5	Avg= 4462.5  RSS=    233 MB  CompCPU=  5.4 sec  Startup= 4452 ms
Avg:	 3975.9	 4500.1	Thr= 4500.1  RSS=    234 MB  CompCPU=  5.8 sec  Startup= 4509 ms
Throughput stats: Avg= 4500.1  StdDev=  139.4  Min= 4270.0  Max= 4650.3  Max/Min=   9% CI95=    2.2% numSamples= 10
Footprint stats:  Avg=  234.2  StdDev=    2.9  Min=  229.9  Max=  239.5  Max/Min=   4% CI95=    0.9% numSamples= 10
Comp CPU stats:   Avg=    5.8  StdDev=    0.8  Min=    5.4  Max=    7.9  Max/Min=  46% CI95=    9.5% numSamples= 10
StartupTime stats:Avg= 4508.8  StdDev=  258.6  Min= 4392.0  Max= 5242.0  Max/Min=  19% CI95=    4.1% numSamples= 10



Results for jdk: /team/lukeli/jdk11Master and opts: -Xmx256m -Xms256m -Xjit:verbose={JITServer},verbose={counts},verbose={compilePerformance},vlog=/team/lukeli/logs/vlog.txt -XX:+UseJITServer -XX:+JITServerUseAOTCache -Xshareclasses:none
Run 0	 3992.8	 4599.0	Avg= 4599.0  RSS=    236 MB  CompCPU=  8.0 sec  Startup= 5134 ms
Run 1	 3967.3	 4596.7	Avg= 4596.7  RSS=    233 MB  CompCPU=  5.8 sec  Startup= 4492 ms
Run 2	 4006.5	 4610.6	Avg= 4610.6  RSS=    236 MB  CompCPU=  5.7 sec  Startup= 4415 ms
Run 3	 4037.7	 4619.2	Avg= 4619.2  RSS=    237 MB  CompCPU=  5.5 sec  Startup= 4529 ms
Run 4	 3965.2	 4234.9	Avg= 4234.9  RSS=    232 MB  CompCPU=  5.5 sec  Startup= 4457 ms
Run 5	 3983.6	 4584.8	Avg= 4584.8  RSS=    234 MB  CompCPU=  5.5 sec  Startup= 4482 ms
Run 6	 4140.9	 4461.0	Avg= 4461.0  RSS=    232 MB  CompCPU=  5.7 sec  Startup= 4435 ms
Run 7	 3995.1	 4480.8	Avg= 4480.8  RSS=    233 MB  CompCPU=  5.5 sec  Startup= 4391 ms
Run 8	 3950.2	 4475.2	Avg= 4475.2  RSS=    230 MB  CompCPU=  5.5 sec  Startup= 4335 ms
Run 9	 3952.7	 4343.6	Avg= 4343.6  RSS=    232 MB  CompCPU=  5.4 sec  Startup= 4373 ms
Avg:	 3999.2	 4500.6	Thr= 4500.6  RSS=    233 MB  CompCPU=  5.8 sec  Startup= 4504 ms
Throughput stats: Avg= 4500.6  StdDev=  129.1  Min= 4234.9  Max= 4619.2  Max/Min=   9% CI95=    2.1% numSamples= 10
Footprint stats:  Avg=  233.5  StdDev=    2.1  Min=  230.4  Max=  237.2  Max/Min=   3% CI95=    0.6% numSamples= 10
Comp CPU stats:   Avg=    5.8  StdDev=    0.8  Min=    5.4  Max=    8.0  Max/Min=  48% CI95=    9.5% numSamples= 10
StartupTime stats:Avg= 4504.3  StdDev=  228.9  Min= 4335.0  Max= 5134.0  Max/Min=  18% CI95=    3.6% numSamples= 10
```

**Comment by mpirvu** [comment]
My 50x grinder on the test jdk_lang_j9 that failed on zlinuxjit passed apparently, though I see one timeout in the console log.
However, the vast majority of the new code should not be exercised without the appropriate cmd line option. Thus, I think this PR is safe to be merged.

File: runtime/oti/j9nonbuilder.h [hunk 21]
```diff
@@ -4291,6 +4291,10 @@ typedef struct J9JITConfig {
 	uint64_t serverUID;
 #endif /* J9VM_OPT_JITSERVER */
 	void (*jitAddNewLowToHighRSSRegion)(const char *name, uint8_t *start, uint32_t size, size_t pageSize);
+#if defined(J9VM_OPT_JITSERVER)
+	void *serverAOTMethodSet;
+	UDATA serverAOTQueryThread;
+#endif // J9VM_OPT_JITSERVER
```
Comment (keithc-ca) [inline runtime/oti/j9nonbuilder.h:4297]: This code is included in C source files where `//` does not legally begin a comment.
This line should instead be:
```c
#endif /* defined(J9VM_OPT_JITSERVER) */
```

**Comment by mpirvu** [comment]
@luke-li-2003 Please see the comment from @keithc-ca and address it in a new PR.

**Comment by keithc-ca** [comment]
I already have a change that fixes that and cleans up other similar misuses.
